Sources/SwiftParser/Declarations.swift

```diff
@@ -311,6 +325,185 @@ extension Parser {
+          placeholder: missingToken(.identifier, text: "decl"),
+          inheritanceClause: nil,
+          genericWhereClause: nil,
+          arena: self.arena
+        )
+      )
+    }
+
+    let attrs = DeclAttributes(
+      attributes: self.parseAttributeList(),
+      modifiers: self.parseDeclModifierList()
+    )
+
+    let recoveryResult: (match: DeclGroupHeaderSyntax.IntroducerOptions, handle: RecoveryConsumptionHandle)?
+    if let atResult = self.at(anyIn: DeclGroupHeaderSyntax.IntroducerOptions.self) {
+      // We are at a keyword that starts a declaration. Parse that declaration.
+      recoveryResult = (atResult.spec, .noRecovery(atResult.handle))
+    } else {
+      // In all other cases, use standard token recovery to find the declaration
+      // to parse.
+      // If we are inside a memberDecl list, we don't want to eat closing braces (which most likely close the outer context)
+      // while recovering to the declaration start.
+      let recoveryPrecedence = inMemberDeclList ? TokenPrecedence.closingBrace : nil
+      recoveryResult = self.canRecoverTo(
+        anyIn: DeclGroupHeaderSyntax.IntroducerOptions.self,
+        overrideRecoveryPrecedence: recoveryPrecedence
+      )
+    }
+
+    if let (match, handle) = recoveryResult {
+      return parseHeaderForDeclarationGroup(
+        attrs: attrs,
+        introducer: match,
+        introducerHandle: handle,
+        allowsMemberBlock: false
+      ).0
+    }
+
+    if inMemberDeclList {
+      if self.currentToken.isEditorPlaceholder {
+        let placeholder = self.parseAnyIdentifier()
+        return RawDeclGroupHeaderSyntax(
+          RawMissingDeclHeaderSyntax(
+            attributes: attrs.attributes,
+            modifiers: attrs.modifiers,
+            placeholder: placeholder,
+            inheritanceClause: nil,
+            genericWhereClause: nil,
+            arena: self.arena
+          )
+        )
+      }
+    }
+
+    return RawDeclGroupHeaderSyntax(
+      RawMissingDeclHeaderSyntax(
+        attributes: attrs.attributes,
+        modifiers: attrs.modifiers,
+        placeholder: missingToken(.identifier, text: "decl"),
+        inheritanceClause: nil,
+        genericWhereClause: nil,
+        arena: self.arena
+      )
+    )
+  }
+
+  /// Parse the header of a declaration group that will belong to some larger construct.
+  ///
+  /// - Parameters:
+  ///   - allowsMemberBlock: If `false`, a member block is definitively *not*
+  ///     expected after this header and if one is found, it should be
+  ///     proactively consumed.
+  mutating func parseHeaderForDeclarationGroup(
+    attrs: DeclAttributes,
+    introducer: DeclGroupHeaderSyntax.IntroducerOptions,
+    introducerHandle: RecoveryConsumptionHandle,
+    allowsMemberBlock: Bool
+  ) -> (RawDeclGroupHeaderSyntax, shouldContinueParsing: Bool) {
+    func eraseToRawDeclGroupHeaderSyntax(
+      _ result: (some RawDeclGroupHeaderSyntaxNodeProtocol, Bool)
```

Comment:
Could you add labels to the tuple elements here? It’s not really obvious what the `Bool` represents here.